### ISSUE TITLE
Problem re-registering on  iOS

### ISSUE THREAD

**Issue by andres-torres-marroquin** [phonegap/phonegap-plugin-push#500]
First of all, thanks for the plugin!

As I have read from a bunch of issues, both closed and open,  the process lifecycle is different on iOS and Android:

Android:
 1. Register
 2. Unregister
 3. Register (gives you a new registration_id)

iOS:
 1. Register
 2. Unregister
 3. Register (gives you the **same** registration_id as step 1)

The problem resides that when I'm trying to send a message from the backend server to GCM to a re-registered iOS device, the GCM server says: NotRegistered.

How can I fix or workaround this?

**Comment by macdonst** [comment]
@andres-torres-marroquin can you send me the parameters you pass into the plugin's init method?

**Comment by andres-torres-marroquin** [comment]
Sure! @macdonst 
```js

window.phonegap_push = PushNotification.init({
  android: {
    senderID: window.GCM_SENDER_ID,
    icon: 'notification_icon',
    forceShow: true
  },
  ios: {
    senderID: window.GCM_SENDER_ID,
    alert: 'true',
    badge: 'true',
    sound: 'true',
    clearBadge: 'true'
  }
});

window.phonegap_push.on('registration', function(data) {
  var device;
  device = new models.Device;
  return device.save({
    registration_id: data.registrationId
  });
});
 ```

**Comment by macdonst** [comment]
Testing development or production push?

**Comment by andres-torres-marroquin** [comment]
I gotta tell you that I'm unsure, but here is a snippet in Python using python-gcm:
```python
from gcm import GCM

API_KEY = 'xxx'
gcm = GCM(API_KEY)

# Iphone
reg_ids = ["xxx"]
registration_id = reg_ids[0]
notification = {
    "body": "body",
    "title": "title",
    "sound": "default",
    
}

response = gcm.json_request(registration_ids=reg_ids, notification=notification, priority='high')

```
Both APNS certificates are setup, and it does work when it's connected the first time, but after `unregister`, I cannot make it work again.

**Comment by fredgalvao** [comment]
Where does the `reg_ids = ["xxx"]` come from? I suspect your issue is on how you handle a new registrationId. My point is: whenever the `registration` event is triggered and the previously saved registrationId is different than the new one received, you need to update your backend/service to make that the new one is the one used by that code when talking to GCM.

**Comment by andres-torres-marroquin** [comment]
@fredgalvao On the start of the app I init the plugin, and save the registrationId on the DB, even if it already is on the DB, it uses a sql unique, so I cannot have more than one row with the same registrationId. In android this works fine, on iOS don't. So I guess the backend is unrelated. Am I missing something?

**Comment by fredgalvao** [comment]
I mean, what does the following scenario outputs:

```ts
window.phonegap_push.on('registration', function(data) {
  var device;
  device = new models.Device;
  return device.save({
    registration_id: data.registrationId
  });
});

// registration triggers with "AAA"
// I expect reg_ids === ["AAA"] on your backend immediately (or a XHR later)

// registration triggers again with "BBB"
// what's the value of reg_ids?
```

**Comment by andres-torres-marroquin** [comment]
This happens on iOS only so here is the debug console output:

```
2016-01-14 08:38:33.624 Buckets[13572:5587173] Push Plugin register called
2016-01-14 08:38:33.624 Buckets[13572:5587173] PushPlugin.register: setting badge to true
2016-01-14 08:38:33.624 Buckets[13572:5587173] PushPlugin.register: clear badge is set to 1
2016-01-14 08:38:33.624 Buckets[13572:5587173] PushPlugin.register: better button setup
2016-01-14 08:38:33.626 Buckets[13572:5587173] GCM Sender ID 918279578126
2016-01-14 08:38:33.626 Buckets[13572:5587173] Using GCM Notification
2016-01-14 08:38:34.780 Buckets[13572:5586678] Push Plugin register success: <5d9ebd2b aeae24eb 1bf62fb0 b56cb306 b34613e3 72a54262 669516c7 72918364>
2016-01-14 08:38:35.585: GGLInstanceID | Invalid last checkin timestamp in future.
2016-01-14 08:38:35.620: GGLInstanceID | Unable to find token in cache Error Domain=com.google.iid Code=-25300 "(null)"
2016-01-14 08:38:37.971 Buckets[13572:5586678] GCM Registration Token: kuzlcZAtkXI:APA91bG3T2FglVipR7r-kJXGK0Rc9I4PqdF4wZAdJTk1s2s66Wf7LIlNWutaM5KVvlkvOwyP-Kfp12Fm7L1oyRnB6tiB0YVda3ftOWVxjmOQUNdvAe3ylzlLZMxAd0HBQaAMnhxfKwHI
2016-01-14 08:38:44.390 Buckets[13572:5586678] active
2016-01-14 08:38:44.391 Buckets[13572:5586678] PushPlugin clearing badge
2016-01-14 08:40:27.386 Buckets[13572:5586678] active
2016-01-14 08:40:27.387 Buckets[13572:5586678] PushPlugin clearing badge
2016-01-14 08:40:40.773 Buckets[13572:5588107] Push Plugin register called
2016-01-14 08:40:40.774 Buckets[13572:5588107] PushPlugin.register: setting badge to true
2016-01-14 08:40:40.774 Buckets[13572:5588107] PushPlugin.register: clear badge is set to 1
2016-01-14 08:40:40.775 Buckets[13572:5588107] PushPlugin.register: better button setup
2016-01-14 08:40:40.776 Buckets[13572:5588107] GCM Sender ID 918279578126
2016-01-14 08:40:40.777 Buckets[13572:5588107] Using GCM Notification
2016-01-14 08:40:40.925 Buckets[13572:5586678] Push Plugin register success: <5d9ebd2b aeae24eb 1bf62fb0 b56cb306 b34613e3 72a54262 669516c7 72918364>
2016-01-14 08:40:40.961 Buckets[13572:5586678] GCM Registration Token: kuzlcZAtkXI:APA91bG3T2FglVipR7r-kJXGK0Rc9I4PqdF4wZAdJTk1s2s66Wf7LIlNWutaM5KVvlkvOwyP-Kfp12Fm7L1oyRnB6tiB0YVda3ftOWVxjmOQUNdvAe3ylzlLZMxAd0HBQaAMnhxfKwHI
```

Between `2016-01-14 08:38:44.391` and `2016-01-14 08:40:27.386` an unregister occurred, when the unregister happened and I tried to push a notification from the backend, GCM says that `kuzlcZAtkXI:APA91bG3T2FglVipR7r-kJXGK0Rc9I4PqdF4wZAdJTk1s2s66Wf7LIlNWutaM5KVvlkvOwyP-Kfp12Fm7L1oyRnB6tiB0YVda3ftOWVxjmOQUNdvAe3ylzlLZMxAd0HBQaAMnhxfKwHI` is Not Registered (which is fine) and the backend it removes them from the DB, when I retry the registration at `2016-01-14 08:40:27.386` it registers successfully, and adds to the DB the "new" registration token (which is the same token as the first one), and when I try to push a notification from the backend, GCM says that is Not Registered and the backend automatically removes it too.

Does that make sense?

**Comment by andres-torres-marroquin** [comment]
Is there something I can do to figure how to fix this? Thanks!

**Comment by Muji86** [comment]
May be stupid but are you certain you are not caching the old ID? I don't understand why you get the same ID after registering.
Another suggestion would be to raise it with Google if you are certain the same ID is coming back second time round.

**Comment by Papipo** [comment]
The same is happening to me, I can see the same token on XCode debugging: `GCM Registration Token: nA1uhx9xUqc:APA91bHFFBCNsl...`

**Comment by dtruel** [comment]
I am having this same issue, after reinstall of app.

**Comment by Alphatiger** [comment]
Hi Facing the same problem, eventually find out the way GCM and APNS works is different.

For GCM unregister call will remove existing reg id and registration call gives new reg id .

For IOS app+device+(same app version) gives the same id  irrespective of the user unregister calls.
So for any user logout and new user login 
olduser and new user will have same regid

This scenario has to be handled by our server code.

If i am wrong please feel free to comment.

Thanks

**Comment by andres-torres-marroquin** [comment]
I basically did the same, for android and ios.

**Comment by voneddy** [comment]
Looks like proper uninstall  (rather than rebuilding via xCode) (iPhone -> settings -> general -> storage... -> manage (local) -> uninstall your app) will generate a new token...

**Comment by Muji86** [comment]
Why do we need a new token anyway? If you unregister the device you won't be able to send a notification to it anyway, when you reregister you will get the same reg id and you will be able to send to it again.
I have completed my push notification bit of the app and tested it with the above result.

**Comment by andres-torres-marroquin** [comment]
@voneddy you get a new token because the app got uninstalled.

**Comment by OscarGalindo** [comment]
@Muji86 what if you want to unsubscribe to a topic and subscribe to a new one?

**Comment by Muji86** [comment]
@OscarGalindo I can't comment on topics as I haven't used them myself but reading about it, it appears that the behaviour would be the same, you still get a single GCM reg id.

**Comment by OscarGalindo** [comment]
yep, i get the same reg id but it's still register to old topics, thats the problem

**Comment by Muji86** [comment]
What is the code to subscribe to a new topic?

**Comment by OscarGalindo** [comment]
it's in the init from push, you add topic: string[] in the options

**Comment by Muji86** [comment]
Yes I understand, so how do you change the topic? the init is called once, right?

**Comment by OscarGalindo** [comment]
yep.. that's the problem, and now i'm trying to unregister and register again but it continues subscribed to the first topic

**Comment by Muji86** [comment]
Are you changing the topic then  building the apk again and installing it on the device?

**Comment by OscarGalindo** [comment]
no! i try to do it on the fly, i want to unsubscribe and subscribe without uninstall nothing

**Comment by vishnuvathsan** [comment]
Facing the same problem except for the part register-unregister-register.
In such situation, I'm getting the same token and was able to send notification to that device for some time (roughly 4 hours). So I assume after such time the token got expired and thus I get Not Registered response from GCM (still I'm getting the same token from the plugin if I do a `PushNotification.init(pushInitConfig)` there). So it seems like the token refresh is not working properly for iOS.

Further, if I do an unregister call and send a push, GCM returns a success response, however, I won't receive any notification on my device.

All these issues are with **iOS devices only**. In Android, it is working as intended.

**Comment by Muji86** [comment]
Are you able to share your code?

**Comment by vishnuvathsan** [comment]
i have created a new issue(#1117) with detail
